### ISSUE TITLE
empty overlap set!

### ISSUE THREAD

**Issue by Adelaam** [isovic/racon#103]
Hello,

I am trying Racon for first time. I am getting this error when I try to polish my data. I have tried with the files provided for the test and I get same error.

[racon::Polisher::initialize] loaded target sequences
[racon::Polisher::initialize] loaded sequences
[racon::Polisher::initialize] error: empty overlap set!

Thank you very much.

**Comment by rvaser** [comment]
Hi,
can you paste the command you are using?

Best regards,
Robert

**Comment by Adelaam** [comment]
'/home/markwilks/racon/build/bin/racon' -t 4 '/home/markwilks/Desktop/M.abscessus/barcode02/barcode02new/barcode02silter2kbq7.fastq.gz' '/home/markwilks/Desktop/M.abscessus/barcode02/barcode02new/barcode02newminimap/barcode02newoverlap.paf.gz' '/home/markwilks/Desktop/M.abscessus/barcode02/barcode02new/barcode02newminimap/barcode02newminimap.contigs.fa' 
[racon::Polisher::initialize] loaded target sequences
[racon::Polisher::initialize] loaded sequences
[racon::Polisher::initialize] error: empty overlap set!
I called the second overlap but is the one I got from minimap2 (reads-assembly).
BW,

**Comment by rvaser** [comment]
Could you please copy the minimap2 command you used to get the .paf file?

**Comment by Adelaam** [comment]
'/home/markwilks/minimap2-2.13_x64-linux/minimap2' -t 4 -map-ont '/home/markwilks/Desktop/M.abscessus/barcode02/barcode02new/barcode02newminimap/barcode02newminimap.contigs.fa' '/home/markwilks/Desktop/M.abscessus/barcode02/barcode02new/barcode02silter2kbq7.fastq.gz' | gzip -1 > '/home/markwilks/Desktop/M.abscessus/barcode02/barcode02new/barcode02newminimap/barcode02newoverlap.paf.gz'

**Comment by Adelaam** [comment]
Yeah, I did a mistake with the output got from minimpa2.

Thank you!

**Comment by apoosakkannu** [comment]
Hi, I am encountering similar problem, empty overlap set. I am very new to the metagenomics as well as racon. My commands are following, 

minimap2 -x map-ont 98ZLc_assembly_NP.fasta 98ZLc_nanopore_filt.fastq > 98ZLc_assembly_minimap_NP.paf

racon 98ZLc_nanopore_filt.fastq 98ZLc_assembly_minimap_NP.paf 98ZLc_nanopore_filt_ava_assembly.fasta > \ 98ZLc_nanopore_filt_racon.fasta

could you please help me to resolve this problem?

Thanks.

**Comment by rvaser** [comment]
Hi,
it appears that you misplaced the reference file, i.e. in the minimap2 command you are mapping reads to `98ZLc_assembly_NP.fasta` and in the racon command you want to polish `98ZLc_nanopore_filt_ava_assembly.fasta`. Those two files should match.

Best regards,
Robert

**Comment by apoosakkannu** [comment]
Thanks for your kind reply. It has worked. Now i am getting slightly different following error,

minimap2 -x sr 98ZLc_nanopore_assembly_racon_metka2.fasta 98ZLc_combineraw_IL.fastq > 98ZLc_mini_IL.paf


racon 98ZLc_nanopore_raw.fastq 98ZLc_mini_IL.paf 98ZLc_nanopore_assembly_racon_metka2.fasta > 98ZLc_nanopore_assembly_racon_metka2_raconIL.fa
[racon::Polisher::initialize] loaded target sequences 1.170 s
[racon::Polisher::initialize] loaded sequences 152.444 s
[racon::Polisher::initialize] loaded overlaps 116.237 s
[racon::Overlap::find_breaking_points] error: overlap is not transmuted! 

Could you please give some insight on this error and any possibility to overcome.

**Comment by rvaser** [comment]
It appears again that you misplaced the files :D You are mapping `98ZLc_combineraw_IL.fastq` to your assembly with minimap2 and in racon you are using different reads `98ZLc_nanopore_raw.fastq`. These two files should also match :)

**Comment by apoosakkannu** [comment]
my bad. Thanks. But again some error so, anbu@bubulin:~/polyplax/downstreamanalysiswithvaclavassembly$ racon -t 19 98ZLc_combineraw_IL.fastq 98ZLc_mini_IL.paf 98ZLc_nanopore_assembly_racon_metka2.fasta > 98ZLc_nanopore_assembly_racon_metka2_raconIL.fa
[racon::Polisher::initialize] loaded target sequences 1.175 s
[racon::Polisher::initialize] loaded sequences 200.535 s
[racon::Overlap::transmute] error: unequal lengths in sequence and overlap file for sequence A00419:60:H7KWLDRXX:2:2101:3522:1016!

**Comment by rvaser** [comment]
Was the Illumina file `98ZLc_combineraw_IL.fastq` created by joining 2 paired end files?

**Comment by apoosakkannu** [comment]
yes, it is !

**Comment by rvaser** [comment]
The error is probably that reads from a pair have equal names up to the first white space. You can add 1 to the first read and 2 to the second read name. Here is a script I wrote a while back: https://github.com/isovic/racon/issues/68#issuecomment-386223150

**Comment by apoosakkannu** [comment]
thanks. so i made script.py file and saved it my working directory and ran the script. got the following error, 
anbu@bubulin:~/polyplax/downstreamanalysiswithvaclavassembly$ python script.py 98ZLc_combineraw_IL.fastq 
Traceback (most recent call last):
  File "script.py", line 46, in <module>
    if (valid):
NameError: name 'valid' is not defined

am i doing something wrong?

**Comment by rvaser** [comment]
Hmmm that is weird, I tried with both python2 and python3 and it works.

**Comment by apoosakkannu** [comment]
i have fastq file, is it ok

**Comment by rvaser** [comment]
Yeah, It should be a fastq file. Did you by any chance mess up the indentation in the script?

**Comment by apoosakkannu** [comment]
i just copied it. ii have not changed anything.

**Comment by apoosakkannu** [comment]
if you have script.py file, could you please attach it. I will try with it.

**Comment by rvaser** [comment]
Just a sec.

**Comment by apoosakkannu** [comment]
Thanks. It is running now. i guess it will run for a while?

**Comment by rvaser** [comment]
If you have a big file it might take a while :)

**Comment by apoosakkannu** [comment]
it is 22.1 gb, i guess big enough.

**Comment by rvaser** [comment]
Indeed it is. I hope you run the command with piping to a new file!

**Comment by apoosakkannu** [comment]
no i did not do, i am new to scripting. could you give me how to do. it is running on the screen itself.

**Comment by rvaser** [comment]
`python rename.py 98ZLc_combineraw_IL.fastq > 98ZLc_renamed_IL.fastq`

**Comment by apoosakkannu** [comment]
thanks.

**Comment by apoosakkannu** [comment]
do i need to do the minimap with the renamed file before polishing with racon?

**Comment by rvaser** [comment]
Unfortunately yes.

**Comment by apoosakkannu** [comment]
great, thanks.

**Comment by apoosakkannu** [comment]
Hi, I used racon to polish my nanopore sequence assembly by mapping the Illumina sequences. As we discussed above i need to rename my illumina sequences. I used the renamed sequences for polishing my nanopore assembly. The process was done without any error in racon. But I used the polished assembly output to make coverage profile using minimap and samtools. Here comes the problem, my scaffolds are not making any coverage profile for both nanopore and illumina. I wonder could it be problem related to renaming the file? Could you give me an idea for solving this problem.

**Comment by rvaser** [comment]
Not sure what could be wrong. Can you paste all the commands you were using?

**Comment by apoosakkannu** [comment]
Please find the following commands,


minimap2 -x sr 98ZLc_nanopore_assembly_racon_metka2.fasta 98ZLc_combineraw__renamed_IL.fastq > 98ZLc_mini_renamed_IL.paf

racon 98ZLc_combineraw_renamed_IL.fastq 98ZLc_mini_renamed_IL.paf 98ZLc_nanopore_assembly_racon_metka2.fasta > 98ZLc_nanopore_assembly_racon_metka2_racon2.fa

**Comment by rvaser** [comment]
Is the file `98ZLc_nanopore_assembly_racon_metka2_racon2.fa` empty by any chance?

**Comment by apoosakkannu** [comment]
no. they seem to be normal as earlier assembly.

**Comment by apoosakkannu** [comment]
I mean size of the file is similar to 98ZLc_nanopore_assembly_racon_metka2.fasta.

**Comment by rvaser** [comment]
Did you run coverage profiles on the unpolished assembly file and it worked? Did you change any parameters in those commands?

**Comment by apoosakkannu** [comment]
Yes, I tried with 98ZLc_nanopore_assembly_racon_metka2.fasta and it worked. I have not changed any parameters.

**Comment by rvaser** [comment]
Are you using different files now? Can you copy these commands as well?

**Comment by apoosakkannu** [comment]
minimap2 -t 19 -ax map-ont 98ZLc_nanopore_assembly_racon_metka2.fasta 98ZLc_nanopore_raw.fastq | samtools view --threads 19 -Sb -F 0x104 - | samtools sort --threads 19 - > np_cov.bam
samtools depth -aa np_cov.bam | awk -F "\t" '{a[$1] += $3; b[$1]++} END{OFS = ","; for (i in a) print i, a[i]/b[i]}' > np_cov.csv

**Comment by rvaser** [comment]
And you just replaced `98ZLc_nanopore_assembly_racon_metka2.fasta` with `98ZLc_nanopore_assembly_racon_metka2_racon2.fa` and it did not work? Any error message? Did you check that the path is alright?

**Comment by apoosakkannu** [comment]
Yes, I just replaced it. There was no error. I am using them in same folder, so path should not be a problem. The output is similar to others. But when i use .csv file in mmgenome2, Rstudio, then i came to know they are not creating the coverage as others.

**Comment by rvaser** [comment]
No idea then :/

**Comment by apoosakkannu** [comment]
ok, thanks for your help. I will check it out.

**Comment by apoosakkannu** [comment]
Hi, I wonder could it be possible to modify your script (rename.zip) to change the name of the fasta file? if possible please give me some info for modification. Thanks in advance.

**Comment by rvaser** [comment]
Do you want to change sequence names or the actual file that is being produced?

**Comment by apoosakkannu** [comment]
I wonder the error problem in creating the coverage file in my polished assembly using illumina could be due to the rename of the illumina sequence names. So i thought if i could change the assembly name also similar to illumina sequences would help me to solve this problem. so basically want to change the name of the assembly file produced by racon+illumina polishing. What do you think?

**Comment by rvaser** [comment]
Not sure that this will help. Is the file `np_cov.bam` empty?

**Comment by apoosakkannu** [comment]
bam file is not empty and all the files have similar size like normal files. thats why i was thinking it could be possibly due to the name change.

**Comment by rvaser** [comment]
Can you run `head -n 1 98ZLc_nanopore_assembly_racon_metka2.fasta` and `head -n 1 98ZLc_nanopore_assembly_racon_metka2_racon2.fa`?

**Comment by apoosakkannu** [comment]
please see the following,

anbu@bubulin:~/polyplax/downstreamanalysiswithvaclavassembly$ head -n 1 98ZLc_nanopore_assembly_racon_metka2.fasta 
>PolyplaxNanoScarf_Scaffold100006:1.0-196.0:1.0-195.0
anbu@bubulin:~/polyplax/downstreamanalysiswithvaclavassembly$ head -n 1 98ZLc_nanopore_assembly_racon_metka2_racon2.fa 
>PolyplaxNanoScarf_Scaffold100006:1.0-196.0:1.0-195.0 LN:i:194 RC:i:618 XC:f:1.000000

**Comment by rvaser** [comment]
No idea what the problem is.

**Comment by apoosakkannu** [comment]
I wonder what is this addition in LN:i:194 RC:i:618 XC:f:1.000000, is it due to the renaming of the file. If so the when i do the following, 

minimap2 -t 19 -ax map-ont 98ZLc_nanopore_assembly_racon_metka2_racon2.fasta 98ZLc_nanopore_raw.fastq | samtools view --threads 19 -Sb -F 0x104 - | samtools sort --threads 19 - > np_cov.bam
samtools depth -aa np_cov.bam | awk -F "\t" '{a[$1] += $3; b[$1]++} END{OFS = ","; for (i in a) print i, a[i]/b[i]}' > np_cov.csv


the names of the input files will be different right?

**Comment by rvaser** [comment]
`LN:i:194 RC:i:618 XC:f:1.000000` are SAM tags produced by racon which denote the contig length, number of reads used for polishing the contig and the percentage of contig windows polished, respectively. You can try removing them manually for a couple of contigs and see if you get coverages.

**Comment by apoosakkannu** [comment]
ok, i will try.

**Comment by Velastegui** [comment]
Hi!
I am trying to find the best assembler and I'm analyzing different methods using long sequences by Nanopore. While trying to get the new file containing the overlaps by: 
```
bwa index r18_guppy_combine.fastq 

bwa mem r18_guppy_combine.fastq r18_guppy_combine.fastq > R18_bwa_mem_racon1.sam 
```
I get a sam file (R18_bwa_mem_racon1.sam) which is not empty: 
```
ls -l
total 9251888
-rw-rw-r-- 1 edgar edgar 3296041316 jun 22 21:48 R18_bwa_mem_racon1.sam
-rw-rw-r-- 7 edgar edgar 3292112286 mei  6 17:39 r18_guppy_combine.fastq
-rw-rw-r-- 1 edgar edgar         20 jun 19 16:25 r18_guppy_combine.fastq.amb
-rw-rw-r-- 1 edgar edgar   25904106 jun 19 16:25 r18_guppy_combine.fastq.ann
-rw-rw-r-- 1 edgar edgar 1634194112 jun 19 16:25 r18_guppy_combine.fastq.bwt
-rw-rw-r-- 1 edgar edgar  408548509 jun 19 16:25 r18_guppy_combine.fastq.pac
-rw-rw-r-- 1 edgar edgar  817097064 jun 19 16:46 r18_guppy_combine.fastq.sa
-rw-rw-r-- 1 edgar edgar          0 jun 24 09:30 R18_racon1.fasta
```
And when I am doing racon says that there is an empty overlap:
```
racon r18_guppy_combine.fastq R18_bwa_mem_racon1.sam r18_guppy_combine.fastq > R18_racon1.fasta 
[racon::Polisher::initialize] loaded target sequences 103.960 s
[racon::Polisher::initialize] loaded sequences 99.487 s
[racon::Polisher::initialize] error: empty overlap set!    
```
The raw reads are from Guppy which I also selected them as the draft genome. 
THe procedure I want to do is to run 4 times this pipeline and finishing with Medaka. 

Hope you could help me out.
THanks in advance.

**Comment by rvaser** [comment]
Hello!
are you trying to correct errors in your reads or polish your assembly? Your commands look like the former (i.e. you are mapping your reads to each other) but I am not sure why all overlaps are filtered out. Can you check that the first alignment record in the SAM has the exact sequence header as the corresponding sequence in the FASTQ file? Also, racon filters out low quality alignments (0.3 error threahold), so you cna chack that too.

If you are actually trying to polish your assembly, you have to map your reads to the assembly and then run racon with

```bash
Racon -t <threads> reads mappings assembly > polished_assembly
```

Best regards,
Robert

**Comment by Velastegui** [comment]
Hi!
Sorry, I guess I was not clear. Indeed, I am trying to correct errors because I don't use any assembly. I already checked and both files (SAM and FASTQ) have the same sequence header. On the other hand, I don't know how to check the low quality alignments. Where can I see the threshold? 

Thanks for your answer!

**Comment by rvaser** [comment]
If SAM file is used, the error rate is calculated as `1 - min(q,t) / max(q,t)`, where q and t represent number of matches and mismatches (obtained from CIGAR) in query and target, respectively. I am not sure if that is the problem though. I would advise you to run the following commands to see if everyrhing works fine:
```bash
minimap2 -t 12 -ax ava-ont --dual=yes r18_guppy_combine.fastq r18_guppy_combine.fastq > overlaps.sam
racon -t 12 -f r18_guppy_combine.fastq overlaps.sam r18_guppy_combine.fastq > r18_polished.fasta
```

Notes:
- one or two iterations of self-error correction should suffice
- you should pass parameter `-f` to racon so that it takes all found overlaps per read (main difference with respect to assembly polishing(
- racon expect dual overlapa (a-b and b-a) in self error-correction so `--dual=yes` is passed to minimap2
-

**Comment by Velastegui** [comment]
Thank you! So far everything is working.

**Comment by Jenny-Tamboli** [comment]
Hi I am new to Recon and was trying to run the test files (racon/test/data) but it gives me error as shown below: 
Code: racon sample_reads.fastq.gz sample_overlaps.paf.gz sample_reference.fasta.gz >racon1.fa

errors: [racon::Polisher::initialize] loaded target sequence 0.001033 s
           [racon::Polisher::initialize] loaded sequences 0.109507 s
           [racon::Polisher::initialize] error: empty overlap set!

Can you tell me what am I doing wrong?

Thanks!

**Comment by rvaser** [comment]
Hello Jenny,
can you please paste the command you used to obtain the `sample_overlaps.paf.gz` file?

Best regards,
Robert

**Comment by Jenny-Tamboli** [comment]
Hi,
It is already present in /test/data folder in racon-1.3.2. 

Thanks.

**Comment by rvaser** [comment]
Ugh I am sorry, misread the inquiry. If you want to run the test data, you have to use the layout file:
```bash
racon sample_reads.fastq.gz sample_overlaps.paf.gz sample_layout.fasta.gz > racon1.fa
```

The reference file is used to assess the accuracy.

Best regards,
Robert

**Comment by Jenny-Tamboli** [comment]
Thank you so much for quick responses. It worked this time!!

Regards,
Jenny

**Comment by kjoost** [comment]
Hi,

I have some issues with running racon. my codes in racon and minimap2 are: minimap2 -x map-ont /home/fronglesquad/ISB_AB-contigs.fasta /home/fronglesquad/snakemake_V2_ISB8A-12-12-19/filtered/nanofilt_out.fastq.gz > aln_ISB_V1.sam
racon --threads 10   /home/fronglesquad/snakemake_V2_ISB8A-12-12-19/filtered/nanofilt_out.fastq.gz /home/fronglesquad/aln_ISB_V1.sam.gz /home/fronglesquad/ISB_AB-contigs.fasta > output_racon
I get the following error message: Illegal instruction (core dumped).
I dont get what I am doing wrong could you maybe help me?
